src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs

```diff
@@ -0,0 +1,1182 @@
+
+            try
+            {
+                int totalBytesRead = 0;
+
+                while (buffer.Length != 0)
+                {
+                    // Sync over async here -- QUIC implementation does it per-I/O already; this is at least more coarse-grained.
+                    if (_dataPayloadRemaining <= 0 && !ReadNextDataFrameAsync(response, CancellationToken.None).GetAwaiter().GetResult())
+                    {
+                        // End of stream.
+                        break;
+                    }
+
+                    if (_buffer.ActiveLength != 0)
+                    {
+                        // Some of the payload is in our receive buffer, so copy it.
+
+                        int copyLen = (int)Math.Min(buffer.Length, _buffer.ActiveLength);
+                        _buffer.ActiveSpan.Slice(copyLen).CopyTo(buffer);
+
+                        totalBytesRead += copyLen;
+                        _dataPayloadRemaining -= copyLen;
+                        _buffer.Discard(copyLen);
+                        buffer = buffer.Slice(copyLen);
+                    }
+                    else
+                    {
+                        // Receive buffer is empty -- bypass it and read directly into user's buffer.
+
+                        int copyLen = (int)Math.Min(buffer.Length, _dataPayloadRemaining);
+                        int bytesRead = _stream.Read(buffer.Slice(copyLen));
+
+                        totalBytesRead += bytesRead;
+                        _dataPayloadRemaining -= bytesRead;
+                        buffer = buffer.Slice(bytesRead);
+                    }
+                }
+
+                return totalBytesRead;
+            }
+            catch (QuicResetException ex)
+            {
+                // Our connection was reset.
+                // This assumes this will be thrown if either remote or local initiated the reset... probably won't hold. TODO.
+
+                Exception abortException = _connection.AbortException;
+                throw new IOException(SR.net_http_client_execution_error, new HttpRequestException(SR.net_http_client_execution_error, abortException ?? ex));
+            }
+            catch (Http3ConnectionException ex)
+            {
+                // A connection-level protocol error has occurred on our stream.
+                _connection.Abort(ex);
+                throw new IOException(SR.net_http_client_execution_error, new HttpRequestException(SR.net_http_client_execution_error, ex));
+            }
+            catch (Exception ex)
+            {
+                _stream.Reset((long)Http3ErrorCode.InternalError);
+                throw new IOException(SR.net_http_client_execution_error, new HttpRequestException(SR.net_http_client_execution_error, ex));
+            }
+        }
+
+        private async ValueTask<int> ReadResponseContentAsync(HttpResponseMessage response, Memory<byte> buffer, CancellationToken cancellationToken)
```

Comment:
Codeshare between ReadResponseContent/ReadResponseContentAsync

Reply:
Any ideas?

Reply:
Maybe I'm mistaken, but the only difference I see is that ReadNextDataFrameAsync is called sync in ReadResponseContent and async in ReadResponseContentAsync. 

At a minimum, you can share the logic to calculate the totalBytesRead and update the buffers.

Reply:
`Stream.Read` vs `Stream.ReadAsync` as well. I can share the exception handling; I'll do that.